openmoc/materialize.py

```diff
@@ -355,3 +360,314 @@ def load_openmc_mgxs_lib(mgxs_lib, geometry=None):
+#          This bias is a result of the use of scalar flux-weighting to compute
+#          MGXS without properly accounting for angular-dependence of the flux.
+# @param mgxs_lib an openmc.mgxs.Library object
+# @param max_fix_src_iters the max number of fixed source iterations per domain
+# @parma max_domain_iters the max number of iterations across domains
+# @param domains the domain to apply SPH factors to ('fissionable' or 'all')
+# @param sph_tol the tolerance on the SPH factor convergence
+# @param fix_src_tol the tolerance on the MOC fixed source calculations
+# @param num_azim number of azimuthal angles to use in fixed source calculations
+# @param track_spacing the track spacing to use in fixed source calculations
+# @param num_threads the number of threads to use in fixed source calculations
+# @return a NumPy array of SPH factors and a new openmc.mgxs.Library object
+#         with the SPH factors applied to each MGXS object
+def compute_sph_factors(mgxs_lib, max_fix_src_iters=10, max_domain_iters=10,
+                        domains='fissionable', sph_tol=1E-6, fix_src_tol=1E-5,
+                        num_azim=4, track_spacing=0.1, num_threads=1, zcoord=0.0):
+
+    import openmc.mgxs
+
+    # For Python 2.X.X
+    if sys.version_info[0] == 2:
+        from compatible import get_openmoc_geometry
+        from process import get_scalar_fluxes
+    # For Python 3.X.X
+    else:
+        from openmoc.compatible import get_openmoc_geometry
+        from openmoc.process import get_scalar_fluxes
+
+    if not isinstance(mgxs_lib, openmc.mgxs.Library):
+        py_printf('ERROR', 'Unable to compute SPH factors with %s which is '
+                  'not an openmc.mgxs.Library', str(openmc.mgxs.Library))
+
+    py_printf('NORMAL', 'Computing SPH factors...')
+
+    # Create an OpenMOC Geometry from the OpenCG Geometry
+    geometry = get_openmoc_geometry(mgxs_lib.opencg_geometry)
+
+    # Load the MGXS library data into the OpenMOC geometry
+    load_openmc_mgxs_lib(mgxs_lib, geometry)
+
+    # Initialize an OpenMOC TrackGenerator
+    geometry.initializeFlatSourceRegions()
+    track_generator = openmoc.TrackGenerator(geometry, num_azim, track_spacing)
+    track_generator.setZCoord(zcoord)
+    track_generator.generateTracks()
+
+    # Initialize an OpenMOC Solver
+    solver = openmoc.CPUSolver(track_generator)
+    solver.setConvergenceThreshold(fix_src_tol)
+    solver.setNumThreads(num_threads)
+    openmc_fluxes = _load_openmc_src(mgxs_lib, solver)
+
+    # Initialize SPH factors
+    num_groups = geometry.getNumEnergyGroups()
+    num_fsrs = geometry.getNumFSRs()
+
+    # Map FSRs to domains to compute domain-averaged fluxes
+    num_domains = len(mgxs_lib.domains)
+    fsrs_to_domains = np.zeros(num_fsrs)
+    for fsr in range(num_fsrs):
+        cell = geometry.findCellContainingFSR(fsr)
+
+        if mgxs_lib.domain_type == 'material':
+            fsrs_to_domains[fsr] = cell.getFillMaterial().getId()
+        else:
```

Comment:
Could someone create a mgxs lib with a `universe` domain (and perhaps in the future mesh domain)? If so, I think it would be worthwhile to check for these other domain types and error out if they are not supported.